huaweicloud/services/cnad/resource_huaweicloud_cnad_advanced_policy.go

```diff
@@ -0,0 +1,277 @@
+				Description: `Whether enable connection protection.`,
+			},
+			"connection_protection_list": {
+				Type:        schema.TypeList,
+				Elem:        &schema.Schema{Type: schema.TypeString},
+				Computed:    true,
+				Description: `The connection protection list.`,
+			},
+			"fingerprint_count": {
+				Type:        schema.TypeInt,
+				Computed:    true,
+				Description: `The fingerprint count.`,
+			},
+			"port_block_count": {
+				Type:        schema.TypeInt,
+				Computed:    true,
+				Description: `The number of port blockages.`,
+			},
+			"watermark_count": {
+				Type:        schema.TypeInt,
+				Computed:    true,
+				Description: `The number of watermarks.`,
+			},
+		},
+	}
+}
+
+func resourceCNADAdvancedPolicyCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	cfg := meta.(*config.Config)
+
+	var (
+		createPolicyHttpUrl = "v1/cnad/policies"
+		createPolicyProduct = "aad"
+	)
+	createPolicyClient, err := cfg.NewServiceClient(createPolicyProduct, "")
+	if err != nil {
+		return diag.Errorf("error creating CNAD Client: %s", err)
+	}
+
+	createPolicyPath := createPolicyClient.Endpoint + createPolicyHttpUrl
+	createPolicyOpt := golangsdk.RequestOpts{
+		MoreHeaders: map[string]string{
+			"Content-Type": "application/json;charset=utf8",
+		},
+		KeepResponseBody: true,
+		JSONBody:         utils.RemoveNil(buildCreatePolicyBodyParams(d)),
+	}
+
+	createPolicyResp, err := createPolicyClient.Request("POST", createPolicyPath, &createPolicyOpt)
+	if err != nil {
+		return diag.Errorf("error creating CNADAdvancedPolicy: %s", err)
+	}
+
+	createPolicyRespBody, err := utils.FlattenResponse(createPolicyResp)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	id, err := jmespath.Search("id", createPolicyRespBody)
+	if err != nil {
+		return diag.Errorf("error creating CNADAdvancedPolicy: ID is not found in API response")
+	}
+	d.SetId(id.(string))
+
+	_, ok1 := d.GetOk("threshold")
+	_, ok2 := d.GetOk("udp")
+	if ok1 || ok2 {
+		if err = updatePolicy(createPolicyClient, d); err != nil {
+			return diag.FromErr(err)
+		}
+	}
+
+	return resourceCNADAdvancedPolicyRead(ctx, d, meta)
+}
+
+func buildCreatePolicyBodyParams(d *schema.ResourceData) map[string]interface{} {
+	bodyParams := map[string]interface{}{
+		"package_id": utils.ValueIngoreEmpty(d.Get("instance_id")),
+		"name":       utils.ValueIngoreEmpty(d.Get("name")),
+	}
+	return bodyParams
+}
+
+func resourceCNADAdvancedPolicyRead(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	cfg := meta.(*config.Config)
+
+	getPolicyClient, err := cfg.NewServiceClient("aad", "")
+	if err != nil {
+		return diag.Errorf("error creating CNAD Client: %s", err)
+	}
+
+	if err := readPolicyDetail(getPolicyClient, d); err != nil {
+		return common.CheckDeletedDiag(d, err, "error retrieving CNADAdvancedPolicy")
+	}
+
+	return nil
+}
+
+func readPolicyDetail(getPolicyClient *golangsdk.ServiceClient, d *schema.ResourceData) error {
```

Comment:
this function will query the policy from HW cloud and set fields to state, it's better to ensure the function only responsible for one thing.